src/merkle_tree.rs

```diff
@@ -0,0 +1,358 @@
+use sha2::{Digest, Sha256};
+
+pub struct MerkleTree {
+    tree: Vec<String>,
+    depth: u8,
+    amount: usize,
+}
+
+impl Default for MerkleTree {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MerkleTree {
+    pub fn new() -> Self {
+        let tree = Vec::new();
+        MerkleTree {
+            tree,
+            depth: 0,
+            amount: 0,
+        }
+    }
+
+    pub fn add_raw(&mut self, raw_text: String) {
+        let mut hasher = Sha256::new();
+        hasher.update(raw_text);
+        let hashed: [u8; 32] = hasher.finalize().into();
+        let hashed_string = hex::encode(hashed);
+
+        self.add(hashed_string);
+    }
+
+    pub fn add(&mut self, hashed_string: String) {
+        self.resize_tree();
+
+        let gap = 2_i8.pow(self.depth as u32) - (self.amount + 1) as i8;
+
+        let non_leaf_nodes = 2_i8.pow(self.depth as u32) as usize - 1;
+        let amount_of_copies = self.tree.len() - self.amount - non_leaf_nodes;
+
+        if gap > 0 && amount_of_copies == 0 {
+            for _ in 0..gap {
+                self.tree.push(hashed_string.clone());
+            }
+            self.tree.push(hashed_string.clone());
+        } else if gap <= 0 {
+            self.tree.pop();
+            self.tree.push(hashed_string);
+        } else if amount_of_copies > 0 {
+            self.tree.pop();
+            self.tree
+                .insert(non_leaf_nodes + self.amount, hashed_string);
+        }
+
+        self.amount += 1;
+        self.rehash_tree(0);
+    }
+
+    fn resize_tree(&mut self) {
+        if self.amount == 0 {
+            self.depth = 1;
+            self.tree.insert(0, "".to_string());
+        }
+        if self.amount == 1 {
+            return;
+        }
+        if MerkleTree::number_is_exact_log_2(self.amount as f32) {
+            self.depth += 1;
+            for _ in 0..self.amount {
+                self.tree.insert(self.amount - 1, "".to_string());
+            }
+        }
+    }
+
+    fn number_is_exact_log_2(num: f32) -> bool {
```

Comment:
```suggestion
    fn is_power_of_two(num: f32) -> bool {
```
I find this to be a more descriptive name.